### ISSUE TITLE
Windows source builds silently compute wrong results in parallel_for kernels: FindOpenMP links both vcomp140 and libomp140 into torch_cpu

### ISSUE THREAD

**Issue by egeboy35** [pytorch/pytorch#193784]
### 🐛 Describe the bug

Windows source builds (MSVC, no MKL) link **two OpenMP runtimes** into `torch_cpu.dll`
— vcomp140 for `#pragma omp` codegen and LLVM's libomp140 for the `omp_*` API calls —
which makes `at::parallel_for` silently execute **every chunk on every worker thread**.
Any parallel kernel with a non-idempotent body then computes garbage, non-deterministically.

User-visible repro (CPU, bf16; fp16 too — this is the path `Muon`'s Newton-Schulz hits,
which is how we found it):

```python
import torch
torch.manual_seed(0)
u = torch.randn(8, 4).bfloat16()
xT = u.T                      # transposed view -> routes to gemm_transa_
xC = xT.contiguous()
m = torch.randn(4, 4).bfloat16()
ref = xC.float() + m.float() @ xT.float()
got = torch.addmm(xC, m, xT, beta=1.0)     # beta != 0 -> non-idempotent body
print((got.float() - ref).abs().max())     # ~1e2..1e8, different every run
torch.set_num_threads(1)                   # single thread -> correct
print((torch.addmm(xC, m, xT, beta=1.0).float() - ref).abs().max())  # ~0.008
```

### Evidence chain

1. `beta=0` is correct, `beta!=0` corrupts; single-thread correct, multi-thread
   non-deterministic (5 runs -> 4 different results). Some cells come out exactly
   `beta*C + 16*dot` on a 16-core machine -> the accumulation ran once per core.
2. Instrumented `gemm_transa_`'s `parallel_for` lambda
   (aten/src/ATen/native/cpu/BlasKernel.cpp:413) and logged, per invocation,
   `omp_get_thread_num() / omp_get_num_threads() / begin / end`. One addmm printed
   **16 identical lines**: `tid=0 team=1 begin=0 end=8` — sixteen workers, each
   believing it is serial, each executing the full range concurrently.
3. `dumpbin /dependents torch_cpu.dll`: imports **both** `VCOMP140.DLL` and
   `libomp140.x86_64.dll`. `dumpbin /imports` shows the fatal split:
   - from VCOMP140: `_vcomp_fork` (region spawning), `_vcomp_set_num_threads`, ...
   - from libomp140: `omp_get_thread_num`, `omp_get_num_threads`, `omp_in_parallel`
   So the pragma spawns a vcomp team (16 threads, unconfigured — torch's
   `set_num_threads(8)` went to libomp), while `invoke_parallel`'s chunking math
   (aten/src/ATen/ParallelOpenMP.h:17) asks libomp, which answers `team=1, tid=0`
   for every worker -> `chunk_size = range`, `begin_tid = 0` -> full-range execution
   on all 16 workers.
4. `ONEDNN_VERBOSE=1` shows oneDNN does not execute this addmm (exonerated), and
   the official MKL wheel is unaffected (MKL path bypasses the ATen kernel).
5. `build.ninja` contains 1676 uses of `-openmp:experimental` and zero
   `-openmp:llvm`; the libomp import therefore comes from the *link* line, not
   from codegen flags.

### Root cause

`cmake/Modules/FindOpenMP.cmake` on MSVC x64 selects the flag
`-openmp:experimental` (vcomp codegen; line 125) but its library-selection logic
finds and links LLVM's `libomp` (`OpenMP_libomp_LIBRARY`), so the `omp_*` API
symbols resolve to libomp140 while `#pragma omp` regions run on vcomp. The
module's own comment block (lines 228-240) quotes libomp's warning that multiple
runtimes "may cause crashes or silently produce incorrect results" — which is
exactly what happens. Notably, MSVC **ARM64** already has a special case forcing
`OpenMP_libomp_LIBRARY` to vcomp (lines 270-278); x64 does not.

Blast radius: any `at::parallel_for` body that is not idempotent (the bf16/fp16
`gemm_transa_` specializations, and plausibly `parallel_reduce`'s per-tid result
slots, which would all collapse onto `results[0]`) on MSVC source builds without
MKL. Everything idempotent (most elementwise TensorIterator loops) silently
"works", which is why this can go unnoticed.

### Proposed fix

Extend the existing ARM64 special case to all MSVC (force vcomp as
`OpenMP_libomp_LIBRARY`, matching the `-openmp:experimental` codegen), or
alternatively switch MSVC to `-openmp:llvm` so codegen matches the linked libomp.
I have the vcomp variant building locally and will send a PR with before/after
`dumpbin` output and the repro above as validation.

### Versions

main (2.15.0a0), MSVC BuildTools 2022 17.14, Windows 11, `USE_CUDA=0`, no MKL
(`mkldnn: True, mkl: False`), CPU: 16-thread x86-64.

*Disclosure: investigated and prepared with AI assistance; I reproduced and
verified every step of the evidence chain on my machine.*

cc @malfet @zklaus @peterjc123 @mszhanyi @skyline75489 @nbcsm @iremyux @Blackhex @nkhasbag-nv @albanD @pragupta

**Comment by egeboy35** [comment]
Follow-up from #193792 (closed pending triage, per process — my mistake sending
the PR before this issue was marked actionable). Summary for triage: the root
cause is proven (`dumpbin` shows `torch_cpu.dll` importing both `VCOMP140` and
`libomp140`; instrumentation shows every worker seeing `team=1/tid=0` and
executing the full range), and the fix — extending the existing MSVC ARM64
vcomp special case in `cmake/Modules/FindOpenMP.cmake` to all MSVC — is
validated locally: a single runtime is imported afterwards, the repro matrix
goes 9/9 correct and bit-identical across runs, and the user-visible symptom
(`torch.optim.Muon` NaN'ing on CPU) disappears. Happy to resend the PR whenever
maintainers confirm the direction.

**Comment by egeboy35** [comment]
This is now `triaged` with `module: correctness (silent)` — is there a direction
call I should wait for before resending the fix?

The fix is small and validated locally (extend the existing MSVC ARM64 vcomp
special case in `cmake/Modules/FindOpenMP.cmake` to all MSVC, so the linked
OpenMP runtime matches the `-openmp:experimental` codegen): after it,
`torch_cpu.dll` imports a single runtime, the bf16/fp16 addmm repro matrix is
9/9 correct and bit-identical across runs, and `torch.optim.Muon` stops NaN'ing
on CPU. I closed the original PR (#193792) per @albanD's process note and can
reopen or send a fresh one whenever this is marked actionable.

**Comment by egeboy35** [comment]
Third and last ping on this one, with an easy out at the end.

This is labelled `triaged` and `module: correctness (silent)` but not `actionable`, and per the
process I have been waiting on that before sending a fix. The fix itself is small and validated
locally: extend the existing MSVC ARM64 vcomp special case in `cmake/Modules/FindOpenMP.cmake` to
all MSVC, so the linked OpenMP runtime matches the `-openmp:experimental` codegen. After it,
`torch_cpu.dll` imports a single runtime, the bf16/fp16 addmm repro matrix is 9/9 correct and
bit-identical across runs, and `torch.optim.Muon` stops producing NaNs on CPU.

What I am asking for is just a direction call, either way:

- if this should be fixed, marking it `actionable` lets me send the PR, or
- if Windows source builds with MSVC OpenMP are not something PyTorch wants to support here, say
  so and I will stop pushing on it.

Happy to hand over the investigation notes to whoever owns this area if that is easier.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

**Comment by frost-intel** [comment]
@albanD @malfet This is a silent correctness, bug, I think this should be actionable.

@egeboy35 I reproduced this independently on a Windows 11 box, but I think your PR has some problems, I'll comment there.

**Comment by zklaus** [comment]
I have investigated the issue and after composed a summary of the findings with Claude:

> Direction is right. The flag and the runtime have to be picked as a pair, and
> there are only two valid pairs:
>
> | codegen | emits | runtime |
> | --- | --- | --- |
> | `/openmp`, `/openmp:experimental` | `_vcomp_*` | vcomp140, or Intel libiomp5md |
> | `/openmp:llvm` | `__kmpc_*` | libomp140 |
>
> We pick `-openmp:experimental`, then `find_library(NAMES omp gomp iomp5)` at
> `FindOpenMP.cmake:312` picks LLVM libomp, which is in the other row. vcomp is
> autolinked anyway because the codegen needs it (the `VCOMP140.DLL` in the
> `dumpbin` output); libomp answers the `omp_*` calls ATen makes directly. Hence
> `team=1, tid=0`.
>
> So extending the ARM64 case at line 270 is right. Notes:
>
> - **Composes with MKL.** `MKL_OPENMP_LIBRARY` is set at line 267, before the
>   `NOT OpenMP_libomp_LIBRARY` guard at 270.
> - **Wheels unaffected.** CD installs `mkl-static` and the smoke test asserts
>   `torch.backends.mkl.is_available()`, so wheels resolve to `libiomp5md` and
>   already exclude vcomp via `Dependencies.cmake:975`. Only non-MKL source builds
>   are hit.
> - **Cannot fix it by widening that guard instead.** Intel documents libiomp5 as
>   "compatible with object files created using the Microsoft* OpenMP run-time
>   library (vcomp)"; LLVM's libomp exports no vcomp entry points at all
>   (`openmp/runtime/src/dllexports`: 725 `__kmpc_*`, zero vcomp). Widening turns
>   a silent wrong answer into unresolved externals.
> - **`-openmp:llvm` is a separate conversation, not this fix.** It is what clang
>   does on MSVC targets (`clang/lib/Driver/ToolChains/MSVC.cpp:316` excludes
>   vcomp). Not blocked by licensing -- MS restricts its own copy under
>   `debug_nonredist` because the switch is experimental, while conda-forge ships
>   `llvm-openmp` for win-64 and we already bundle its macOS build
>   (`third_party/llvm-openmp/LICENSE.txt`). Blocked by scope: new runtime for
>   every MSVC build, a vendored DLL and licence entry in the wheels, MKL pairing
>   to recheck. For no feature we use -- our entire OpenMP surface is `parallel`,
>   `for schedule(static)`, `barrier`, `parallel for if(...)`, all OpenMP 2.0.
>
> Related, all open, all the same decoupling: #113490 (Nov 2023, MKL side),
> #148674 (Mar 2025, `caffe2::mkl` half of the same exclusion, with an unanswered
> Apr 2025 review question asking whether `FindOpenMP.cmake` already covers it).
> Worth settling together or declaring independent.
>
> I have not reviewed the patch and cannot speak to @frost-intel's concerns with
> it; the direction looks right.

**Comment by egeboy35** [comment]
@frost-intel @zklaus thank you both - I had stopped pinging this, so I only just
saw your reproduction and your analysis. Between you that is two independent
confirmations, and I have a third: I hit this again today from the other end,
while chasing what looked like an SDPA bug, on a fresh Windows 11 box with MSVC
14.44.35207 and no MKL. Same import split in `torch_cpu.dll`:

```
VCOMP140.DLL          -> _vcomp_fork, _vcomp_barrier, _vcomp_for_static_end,
                         _vcomp_for_static_simple_init_i8, _vcomp_set_num_threads
libomp140.x86_64.dll  -> omp_get_num_threads, omp_get_thread_num, omp_in_parallel,
                         omp_get_max_threads, omp_set_num_threads
```

No `__kmpc_*` imports at all, and `build.ninja` carries `-openmp:experimental`
and nothing else.

One new piece of evidence that I think closes the "this is just parallel
nondeterminism" reading for good. Integer addition is exactly associative, so no
chunking or reordering can change an integer sum:

```python
>>> y = torch.ones(40_000_000, dtype=torch.int64)
>>> sorted({y.sum().item() for _ in range(8)})
[600000000, 640000000]          # exactly 15x and 40000000 * 16
>>> torch.set_num_threads(1); torch.ones(40_000_000, dtype=torch.int64).sum().item()
40000000
```

Exact multiples of the true answer, from an exactly associative operation. Every
worker summed the whole range.

@frost-intel both your review points on #193792 look right to me - gate on
`CMAKE_${LANG}_COMPILER_ID STREQUAL "MSVC"` rather than `MSVC`, and leave the
library list empty rather than naming `vcomp.lib`. The one thing I want to
measure before resending is the concern in `cmake/public/mkldnn.cmake:20-30`:
with the list empty, `caffe2::openmp` carries no library, and that comment says
binaries pulling dnnl objects then fail to resolve `omp_*`. I have a box that
reproduces this and a working `BUILD_TEST=1` setup, so I can build the C++ test
executables against the patch and report what actually happens.

I will do that regardless. The only thing I am still waiting on is whether this
can be marked `actionable` so the PR is not closed on process again.

Developed with AI assistance (Claude); I measured each result above.

**Comment by egeboy35** [comment]
I built and measured @frost-intel's suggestion rather than the version I
originally sent. It is one condition on the generic library search, which is the
block that actually fires on MSVC x64 - the existing special case above it is
gated on ARM64, so x64 falls through to `NAMES omp gomp iomp5` and finds the
`libomp.lib` shipped in the MSVC toolchain:

```cmake
if ((NOT OpenMP_libomp_LIBRARY) AND (NOT CMAKE_${LANG}_COMPILER_ID STREQUAL "MSVC"))
```

Measured on Windows 11, MSVC 14.44.35207, no MKL, 8 threads:

| | before | after |
|---|---|---|
| `dumpbin -imports torch_cpu.dll` | `VCOMP140.DLL` + `libomp140.x86_64.dll` | `VCOMP140.DLL` only, zero `__kmpc_*` |
| `torch.ones(40M, int64).sum()` x8 | `600000000` / `640000000` | `40000000` every time |
| `torch.arange(1e6, float64).sum()` | wrong | `499999500000.0` |
| SDPA vs MATH, 15 seeds | 9 deviating, 12 nondeterministic | 0 and 0 |

On the concern in `cmake/public/mkldnn.cmake:20-30` - that leaving the list empty
makes `caffe2::openmp` carry no library, so binaries pulling dnnl objects fail to
resolve `omp_*`. I deleted all 125 C++ test executables and forced a fresh link:
**125/125 linked with no unresolved externals**, including `dlconvertor_test`,
which is the only test target linking dnnl directly. It runs and passes, as does
`test_api`. So at least in this configuration the concern does not materialise.

One thing worth knowing for whoever reviews it: the patch changes no compiler
flag - `-openmp:experimental` is used before and after - so nothing recompiles.
The full `pip install -e . -v --no-build-isolation` took 15 minutes, not the
hours a flag change would cost. Validating this is cheap.

That leaves the label. @albanD closed #193792 asking that this be discussed on
the issue and marked `actionable` first. It has since been discussed, reproduced
independently twice, and now measured. Could it be marked `actionable` so I can
resend the PR without it closing on the same grounds?

Developed with AI assistance (Claude); I measured each number above.

**Comment by zklaus** [comment]
I have marked the issue as actionable after the discussion here and also offline discussion with @atalman. Thanks for coming back to this, @egeboy35. Please update and reopen your PR; also take into account #196730, which I think is basically a copy of your PR.

**Comment by egeboy35** [comment]
A note for anyone rebuilding to try the fix in #193792, drafted with Claude:

> The fix does not take effect in an existing build directory.
> `OpenMP_<LANG>_FLAGS`, `OpenMP_<LANG>_LIB_NAMES` and `OpenMP_libomp_LIBRARY` are
> cached: detection is skipped while the first two are set, and even when it
> reruns, a cached libomp path is reused. Delete the build directory, or clear the
> entries first:
>
> ```
> cmake -U "OpenMP_*" <build-dir>
> ```
>
> To check the result, `dumpbin -imports torch_cpu.dll` should list
> `VCOMP140.DLL` and not `libomp140.x86_64.dll`.
